{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import random\n",
    "import sys\n",
    "import os\n",
    "import scipy\n",
    "import math\n",
    "from sklearn.metrics import r2_score\n",
    "from pathlib import Path\n",
    "import subprocess\n",
    "\n",
    "from Bio.SeqIO import read, parse\n",
    "import orthoani\n",
    "\n",
    "from multiprocess import Pool\n",
    "import pandas as pd\n",
    "\n",
    "max_pool = 16\n",
    "lim_file = 200\n",
    "\n",
    "sys.path.append('../')\n",
    "sys.path.append('../../')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from db_sketching.kmer_set import KMerSet, FracMinHash, TruncatedKMerSet, MeanFracMinHash"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def cond(kmer_hash):\n",
    "    hash = (976369 * kmer_hash + 1982627) % 10000\n",
    "    if hash < 50:\n",
    "        return True\n",
    "    else:\n",
    "        return False\n",
    "    \n",
    "\n",
    "def generate_polycond(degree, mod, c, seed_val):\n",
    "    rng = np.random.default_rng(seed_val)\n",
    "    coeffs = np.random.randint(\n",
    "        low = 0,\n",
    "        high = mod,\n",
    "        size = degree+1\n",
    "    )\n",
    "\n",
    "    def poly_cond(kmer_hash):\n",
    "        hash_mod = kmer_hash % mod\n",
    "        end_val = 0\n",
    "        for coeff in reversed(coeffs):\n",
    "            end_val *= hash_mod\n",
    "            end_val += coeff\n",
    "            end_val %= mod\n",
    "\n",
    "        return end_val < (mod / c)\n",
    "    \n",
    "    return poly_cond\n",
    "\n",
    "def all_cond(kmer_hash):\n",
    "    return True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "ANALYSIS_TYPE = \"analysis_type\"\n",
    "ANALYSIS_DATA_DIR = \"analysis_data_dir\"\n",
    "GENOME_FILE_1 = \"genome_file_1\"\n",
    "GENOME_FILE_2 = \"genome_file_2\"\n",
    "ORTHOANI_VAL = \"orthoani_val\"\n",
    "KMER_LENGTH = \"kmer_length\"\n",
    "C_VAL = \"c_val\"\n",
    "CANONICAL = \"canonical\"\n",
    "KMER_ANI_VAL = \"kmer_ani_val\"\n",
    "PEARSON_COEFF = \"pearson_coeff\"\n",
    "PEARSON_COEFF_PVAL = \"pearson_coeff_pval\"\n",
    "SPEARMAN_COEFF = \"spearman_coeff\"\n",
    "SPEARMAN_COEFF_PVAL = \"spearman_coeff_pval\"\n",
    "R2_VAL_RAW = \"r2_val_raw\"\n",
    "R2_VAL_LIN = \"r2_val_lin\"\n",
    "DROPPED_ZEROS = \"dropped_zeros\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_home_dir = \"../../data_temp\"\n",
    "\n",
    "species_analysis_type = \"Single-Species\"\n",
    "species_analysis_data_dirs = {\n",
    "    \"Escherichia coli\": (10,100),\n",
    "    \"Lactobacillus helveticus\": (10,100),\n",
    "    \"Staphylococcus hominis\": (10,100),\n",
    "    \"Mycoplasmoides pneumoniae\": (10,100),\n",
    "    \"Brucella melitensis\": (10,100),\n",
    "    \"Xanthomonas oryzae\": (10,100)\n",
    "}\n",
    "\n",
    "genus_analysis_type = \"Single-Genus\"\n",
    "genus_analysis_data_dirs = {\n",
    "    \"Pectobacterium\": (10,35),\n",
    "    \"Morganella\": (10,35),\n",
    "    \"Xylella\": (10,35),\n",
    "}\n",
    "\n",
    "family_analysis_type = \"Single-Family\"\n",
    "family_analysis_data_dirs = {\n",
    "    \"Enterobacteriaceae\": (10,35),\n",
    "    \"Cyanobiaceae\": (10,35),\n",
    "    \"Rhizobiaceae\": (10,35)\n",
    "}\n",
    "\n",
    "family_genus_analysis_type = \"Single-Family-Multi-Genus\"\n",
    "family_genus_analysis_data_dirs = {\n",
    "    \"31989\": (10,35),\n",
    "    \"49546\": (10,35),\n",
    "    \"186803\": (10,35),\n",
    "    \"186817\": (10,35),\n",
    "}\n",
    "\n",
    "genus_species_analysis_type = \"Single-Genus-Multi-Species\"\n",
    "genus_species_analysis_data_dirs = {\n",
    "    \"Psychrobacter\": (10,35),\n",
    "    \"Aeromonas\": (10,35),\n",
    "    \"Rathayibacter\": (10,35),\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "analysis_types = {\n",
    "    species_analysis_type: species_analysis_data_dirs,\n",
    "    genus_analysis_type: genus_analysis_data_dirs,\n",
    "    family_analysis_type: family_analysis_data_dirs,\n",
    "    family_genus_analysis_type: family_genus_analysis_data_dirs,\n",
    "    genus_species_analysis_type: genus_species_analysis_data_dirs\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "c_val_range = (200,)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Helper function to obtain genome files\n",
    "def get_genome_files(data_home_dir, analysis_type, analysis_data_dir):\n",
    "    genome_files = []\n",
    "    collection_path = os.path.join(data_home_dir,analysis_type,analysis_data_dir)\n",
    "    for filename in os.listdir(collection_path):\n",
    "        full_filename = os.path.join(collection_path,filename)\n",
    "        genome_files.append(full_filename)\n",
    "    return genome_files\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_genome_files(genome_files):\n",
    "    checked_genome_files = []\n",
    "    for file in genome_files:\n",
    "        try:\n",
    "            parsed_file = parse(file,\"fasta\")\n",
    "            assert(len([record for record in parsed_file]) > 0)\n",
    "            checked_genome_files.append(file)\n",
    "        except:\n",
    "            print(f\"File {file} is damaged / invalid\")\n",
    "\n",
    "    return checked_genome_files\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_genome_length(genome_file):\n",
    "    genome_kmer = KMerSet(3)\n",
    "    genome_kmer.insert_file(genome_file)\n",
    "    return genome_kmer.length\n",
    "\n",
    "def compute_length_parallel(genome_files):\n",
    "    with Pool(max_pool) as p:\n",
    "        return p.starmap(get_genome_length,([(g_file,) for g_file in genome_files]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_ortho_ani(genome_file_1, genome_file_2):\n",
    "    try:\n",
    "        genome_1_read = parse(genome_file_1,\"fasta\")\n",
    "        genome_2_read = parse(genome_file_2,\"fasta\")\n",
    "        ortho_ani_value = orthoani.orthoani(genome_1_read,genome_2_read)\n",
    "        return ortho_ani_value\n",
    "    except:\n",
    "        return 0\n",
    "\n",
    "def compute_ortho_ani_parallel(genome_files_1,genome_files_2):\n",
    "    args = [(g1,g2) for g1,g2 in zip(genome_files_1,genome_files_2)]\n",
    "    with Pool(max_pool) as p:\n",
    "        return p.starmap(compute_ortho_ani,args)\n",
    "\n",
    "def compute_pairwise_ortho(genome_files):\n",
    "    genome_files_1, genome_files_2 = zip(*[(g1,g2) for g1 in genome_files for g2 in genome_files])\n",
    "    return compute_ortho_ani_parallel(genome_files_1,genome_files_2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_kmer_sketches(genome_file, kmer_class, sketching_condition, kmer_length, canonical):\n",
    "    genome_kmer = kmer_class(sketching_condition, kmer_length, canonical)\n",
    "    genome_kmer.insert_file(genome_file)\n",
    "    return genome_kmer\n",
    "\n",
    "def compute_kmer_sketches_parallel(genome_files, kmer_length, kmer_class, sketching_cond, canonical):\n",
    "    args = [(g,kmer_class,sketching_cond,kmer_length,canonical) for g in genome_files]\n",
    "    with Pool(max_pool) as p:\n",
    "        return p.starmap(compute_kmer_sketches,args)\n",
    "\n",
    "\n",
    "def compute_kmer_ani(genome_1_kmer, genome_2_kmer):    \n",
    "    kmer_estimated_ani = genome_1_kmer.ANI_estimation(genome_2_kmer)\n",
    "    return kmer_estimated_ani\n",
    "\n",
    "\n",
    "def compute_kmer_ani_parallel(kmer_sketches_1,kmer_sketches_2):\n",
    "    args = [(s1,s2) for s1,s2 in zip(kmer_sketches_1,kmer_sketches_2)]\n",
    "    with Pool(max_pool) as p:\n",
    "        return p.starmap(compute_kmer_ani,args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                         genome_file_1  \\\n",
      "0       ../../data_temp/Single-Genus/Salmonella/64.fna   \n",
      "1       ../../data_temp/Single-Genus/Salmonella/32.fna   \n",
      "2       ../../data_temp/Single-Genus/Salmonella/99.fna   \n",
      "3       ../../data_temp/Single-Genus/Salmonella/43.fna   \n",
      "4        ../../data_temp/Single-Genus/Salmonella/7.fna   \n",
      "..                                                 ...   \n",
      "494  ../../data_temp/Single-Family/Rhizobiaceae/441...   \n",
      "495  ../../data_temp/Single-Family/Rhizobiaceae/292...   \n",
      "496  ../../data_temp/Single-Family/Rhizobiaceae/83.fna   \n",
      "497  ../../data_temp/Single-Family/Rhizobiaceae/129...   \n",
      "498  ../../data_temp/Single-Family/Rhizobiaceae/291...   \n",
      "\n",
      "                                         genome_file_2  orthoani_val  \\\n",
      "0       ../../data_temp/Single-Genus/Salmonella/32.fna      0.985312   \n",
      "1       ../../data_temp/Single-Genus/Salmonella/99.fna      0.984921   \n",
      "2       ../../data_temp/Single-Genus/Salmonella/43.fna      0.983598   \n",
      "3        ../../data_temp/Single-Genus/Salmonella/7.fna      0.984154   \n",
      "4       ../../data_temp/Single-Genus/Salmonella/98.fna      0.986997   \n",
      "..                                                 ...           ...   \n",
      "494  ../../data_temp/Single-Family/Rhizobiaceae/292...      0.780533   \n",
      "495  ../../data_temp/Single-Family/Rhizobiaceae/83.fna      0.781226   \n",
      "496  ../../data_temp/Single-Family/Rhizobiaceae/129...      0.994015   \n",
      "497  ../../data_temp/Single-Family/Rhizobiaceae/291...      0.997140   \n",
      "498  ../../data_temp/Single-Family/Rhizobiaceae/110...      0.997200   \n",
      "\n",
      "     analysis_type analysis_data_dir  \n",
      "0              NaN               NaN  \n",
      "1              NaN               NaN  \n",
      "2              NaN               NaN  \n",
      "3              NaN               NaN  \n",
      "4              NaN               NaN  \n",
      "..             ...               ...  \n",
      "494  Single-Family      Rhizobiaceae  \n",
      "495  Single-Family      Rhizobiaceae  \n",
      "496  Single-Family      Rhizobiaceae  \n",
      "497  Single-Family      Rhizobiaceae  \n",
      "498  Single-Family      Rhizobiaceae  \n",
      "\n",
      "[2586 rows x 5 columns]\n"
     ]
    }
   ],
   "source": [
    "ortho_ani_filename = \"../../data_temp/ortho_ani_values.csv\"\n",
    "ortho_ani_dataframe = pd.read_csv(ortho_ani_filename,index_col=0)\n",
    "print(ortho_ani_dataframe)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                         genome_file_1  \\\n",
      "0    ../../data_temp/Single-Species/Escherichia col...   \n",
      "1    ../../data_temp/Single-Species/Escherichia col...   \n",
      "2    ../../data_temp/Single-Species/Escherichia col...   \n",
      "3    ../../data_temp/Single-Species/Escherichia col...   \n",
      "4    ../../data_temp/Single-Species/Escherichia col...   \n",
      "..                                                 ...   \n",
      "494  ../../data_temp/Single-Family/Rhizobiaceae/441...   \n",
      "495  ../../data_temp/Single-Family/Rhizobiaceae/292...   \n",
      "496  ../../data_temp/Single-Family/Rhizobiaceae/83.fna   \n",
      "497  ../../data_temp/Single-Family/Rhizobiaceae/129...   \n",
      "498  ../../data_temp/Single-Family/Rhizobiaceae/291...   \n",
      "\n",
      "                                         genome_file_2  kmer_length  c_val  \\\n",
      "0    ../../data_temp/Single-Species/Escherichia col...           10     20   \n",
      "1    ../../data_temp/Single-Species/Escherichia col...           10     20   \n",
      "2    ../../data_temp/Single-Species/Escherichia col...           10     20   \n",
      "3    ../../data_temp/Single-Species/Escherichia col...           10     20   \n",
      "4    ../../data_temp/Single-Species/Escherichia col...           10     20   \n",
      "..                                                 ...          ...    ...   \n",
      "494  ../../data_temp/Single-Family/Rhizobiaceae/292...           35    200   \n",
      "495  ../../data_temp/Single-Family/Rhizobiaceae/83.fna           35    200   \n",
      "496  ../../data_temp/Single-Family/Rhizobiaceae/129...           35    200   \n",
      "497  ../../data_temp/Single-Family/Rhizobiaceae/291...           35    200   \n",
      "498  ../../data_temp/Single-Family/Rhizobiaceae/110...           35    200   \n",
      "\n",
      "     canonical  kmer_ani_val  analysis_type analysis_data_dir  \n",
      "0         True      0.996998            NaN               NaN  \n",
      "1         True      0.997966            NaN               NaN  \n",
      "2         True      0.999697            NaN               NaN  \n",
      "3         True      0.999718            NaN               NaN  \n",
      "4         True      0.999652            NaN               NaN  \n",
      "..         ...           ...            ...               ...  \n",
      "494       True      0.849250  Single-Family      Rhizobiaceae  \n",
      "495       True      0.840393  Single-Family      Rhizobiaceae  \n",
      "496       True      0.995491  Single-Family      Rhizobiaceae  \n",
      "497       True      0.997652  Single-Family      Rhizobiaceae  \n",
      "498       True      0.997943  Single-Family      Rhizobiaceae  \n",
      "\n",
      "[142654 rows x 8 columns]\n"
     ]
    }
   ],
   "source": [
    "kmer_sketch_ani_filename = \"../../data_temp/kmer_sketch_ani_values.csv\"\n",
    "kmer_sketch_dataframe = pd.read_csv(kmer_sketch_ani_filename,index_col=0)\n",
    "print(kmer_sketch_dataframe)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    kmer_length  c_val  canonical  pearson_coeff  pearson_coeff_pval  \\\n",
      "0            10     20       True       0.907790        2.218581e-38   \n",
      "0            10     20      False       0.421223        1.413929e-05   \n",
      "0            10     50       True       0.911519        3.280371e-39   \n",
      "0            10     50      False       0.425088        1.155857e-05   \n",
      "0            10    100       True       0.916422        2.327374e-40   \n",
      "..          ...    ...        ...            ...                 ...   \n",
      "0            31    200       True       0.997971        0.000000e+00   \n",
      "0            32    200       True       0.998233        0.000000e+00   \n",
      "0            33    200       True       0.998213        0.000000e+00   \n",
      "0            34    200       True       0.998040        0.000000e+00   \n",
      "0            35    200       True       0.997898        0.000000e+00   \n",
      "\n",
      "    spearman_coeff  spearman_coeff_pval   analysis_type analysis_data_dir  \\\n",
      "0         0.866259         5.424778e-31  Single-Species  Escherichia coli   \n",
      "0         0.482028         4.377992e-07  Single-Species  Escherichia coli   \n",
      "0         0.872789         5.627322e-32  Single-Species  Escherichia coli   \n",
      "0         0.478169         5.570930e-07  Single-Species  Escherichia coli   \n",
      "0         0.869019         2.112540e-31  Single-Species  Escherichia coli   \n",
      "..             ...                  ...             ...               ...   \n",
      "0         0.972592        4.972939e-317   Single-Family      Rhizobiaceae   \n",
      "0         0.972661        2.666582e-317   Single-Family      Rhizobiaceae   \n",
      "0         0.973296        8.426784e-320   Single-Family      Rhizobiaceae   \n",
      "0         0.973959        1.778636e-322   Single-Family      Rhizobiaceae   \n",
      "0         0.974057        6.916919e-323   Single-Family      Rhizobiaceae   \n",
      "\n",
      "    r2_val  dropped_zeros  r2_val_raw  r2_val_lin  \n",
      "0      NaN            NaN         NaN         NaN  \n",
      "0      NaN            NaN         NaN         NaN  \n",
      "0      NaN            NaN         NaN         NaN  \n",
      "0      NaN            NaN         NaN         NaN  \n",
      "0      NaN            NaN         NaN         NaN  \n",
      "..     ...            ...         ...         ...  \n",
      "0      NaN            0.0    0.862895    0.864895  \n",
      "0      NaN            0.0    0.851731    0.853408  \n",
      "0      NaN            0.0    0.844561    0.846218  \n",
      "0      NaN            0.0    0.834326    0.836093  \n",
      "0      NaN            0.0    0.839144    0.841070  \n",
      "\n",
      "[1146 rows x 13 columns]\n"
     ]
    }
   ],
   "source": [
    "corrleation_filename = \"../../data_temp/correlation_values.csv\"\n",
    "correlation_dataframe = pd.read_csv(corrleation_filename,index_col=0)\n",
    "print(correlation_dataframe)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "kmer_class = FracMinHash"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "analysis_type='Single-Genus', analysis_data_dir='Pectobacterium'\n",
      "Using genome files ['../../data_temp/Single-Genus/Pectobacterium/64.fna', '../../data_temp/Single-Genus/Pectobacterium/32.fna', '../../data_temp/Single-Genus/Pectobacterium/99.fna', '../../data_temp/Single-Genus/Pectobacterium/43.fna', '../../data_temp/Single-Genus/Pectobacterium/7.fna', '../../data_temp/Single-Genus/Pectobacterium/98.fna', '../../data_temp/Single-Genus/Pectobacterium/58.fna', '../../data_temp/Single-Genus/Pectobacterium/90.fna', '../../data_temp/Single-Genus/Pectobacterium/14.fna', '../../data_temp/Single-Genus/Pectobacterium/41.fna', '../../data_temp/Single-Genus/Pectobacterium/67.fna', '../../data_temp/Single-Genus/Pectobacterium/73.fna', '../../data_temp/Single-Genus/Pectobacterium/17.fna', '../../data_temp/Single-Genus/Pectobacterium/56.fna', '../../data_temp/Single-Genus/Pectobacterium/50.fna', '../../data_temp/Single-Genus/Pectobacterium/2.fna', '../../data_temp/Single-Genus/Pectobacterium/46.fna', '../../data_temp/Single-Genus/Pectobacterium/52.fna', '../../data_temp/Single-Genus/Pectobacterium/65.fna', '../../data_temp/Single-Genus/Pectobacterium/95.fna', '../../data_temp/Single-Genus/Pectobacterium/45.fna', '../../data_temp/Single-Genus/Pectobacterium/74.fna', '../../data_temp/Single-Genus/Pectobacterium/44.fna', '../../data_temp/Single-Genus/Pectobacterium/12.fna', '../../data_temp/Single-Genus/Pectobacterium/94.fna', '../../data_temp/Single-Genus/Pectobacterium/25.fna', '../../data_temp/Single-Genus/Pectobacterium/60.fna', '../../data_temp/Single-Genus/Pectobacterium/39.fna', '../../data_temp/Single-Genus/Pectobacterium/22.fna', '../../data_temp/Single-Genus/Pectobacterium/48.fna', '../../data_temp/Single-Genus/Pectobacterium/88.fna', '../../data_temp/Single-Genus/Pectobacterium/24.fna', '../../data_temp/Single-Genus/Pectobacterium/9.fna', '../../data_temp/Single-Genus/Pectobacterium/86.fna', '../../data_temp/Single-Genus/Pectobacterium/5.fna', '../../data_temp/Single-Genus/Pectobacterium/89.fna', '../../data_temp/Single-Genus/Pectobacterium/71.fna', '../../data_temp/Single-Genus/Pectobacterium/54.fna', '../../data_temp/Single-Genus/Pectobacterium/23.fna', '../../data_temp/Single-Genus/Pectobacterium/49.fna', '../../data_temp/Single-Genus/Pectobacterium/40.fna', '../../data_temp/Single-Genus/Pectobacterium/18.fna', '../../data_temp/Single-Genus/Pectobacterium/4.fna', '../../data_temp/Single-Genus/Pectobacterium/20.fna', '../../data_temp/Single-Genus/Pectobacterium/63.fna', '../../data_temp/Single-Genus/Pectobacterium/33.fna', '../../data_temp/Single-Genus/Pectobacterium/87.fna', '../../data_temp/Single-Genus/Pectobacterium/80.fna', '../../data_temp/Single-Genus/Pectobacterium/26.fna', '../../data_temp/Single-Genus/Pectobacterium/1.fna', '../../data_temp/Single-Genus/Pectobacterium/47.fna', '../../data_temp/Single-Genus/Pectobacterium/92.fna', '../../data_temp/Single-Genus/Pectobacterium/51.fna', '../../data_temp/Single-Genus/Pectobacterium/37.fna', '../../data_temp/Single-Genus/Pectobacterium/84.fna', '../../data_temp/Single-Genus/Pectobacterium/53.fna', '../../data_temp/Single-Genus/Pectobacterium/70.fna', '../../data_temp/Single-Genus/Pectobacterium/42.fna', '../../data_temp/Single-Genus/Pectobacterium/57.fna', '../../data_temp/Single-Genus/Pectobacterium/72.fna', '../../data_temp/Single-Genus/Pectobacterium/91.fna', '../../data_temp/Single-Genus/Pectobacterium/27.fna', '../../data_temp/Single-Genus/Pectobacterium/19.fna', '../../data_temp/Single-Genus/Pectobacterium/66.fna', '../../data_temp/Single-Genus/Pectobacterium/13.fna', '../../data_temp/Single-Genus/Pectobacterium/16.fna', '../../data_temp/Single-Genus/Pectobacterium/29.fna', '../../data_temp/Single-Genus/Pectobacterium/8.fna', '../../data_temp/Single-Genus/Pectobacterium/21.fna', '../../data_temp/Single-Genus/Pectobacterium/3.fna', '../../data_temp/Single-Genus/Pectobacterium/10.fna', '../../data_temp/Single-Genus/Pectobacterium/30.fna', '../../data_temp/Single-Genus/Pectobacterium/38.fna', '../../data_temp/Single-Genus/Pectobacterium/93.fna', '../../data_temp/Single-Genus/Pectobacterium/78.fna', '../../data_temp/Single-Genus/Pectobacterium/77.fna', '../../data_temp/Single-Genus/Pectobacterium/75.fna', '../../data_temp/Single-Genus/Pectobacterium/35.fna', '../../data_temp/Single-Genus/Pectobacterium/31.fna', '../../data_temp/Single-Genus/Pectobacterium/28.fna', '../../data_temp/Single-Genus/Pectobacterium/82.fna', '../../data_temp/Single-Genus/Pectobacterium/59.fna', '../../data_temp/Single-Genus/Pectobacterium/69.fna', '../../data_temp/Single-Genus/Pectobacterium/62.fna', '../../data_temp/Single-Genus/Pectobacterium/85.fna', '../../data_temp/Single-Genus/Pectobacterium/0.fna', '../../data_temp/Single-Genus/Pectobacterium/79.fna', '../../data_temp/Single-Genus/Pectobacterium/11.fna', '../../data_temp/Single-Genus/Pectobacterium/34.fna', '../../data_temp/Single-Genus/Pectobacterium/6.fna', '../../data_temp/Single-Genus/Pectobacterium/68.fna', '../../data_temp/Single-Genus/Pectobacterium/97.fna', '../../data_temp/Single-Genus/Pectobacterium/15.fna', '../../data_temp/Single-Genus/Pectobacterium/55.fna', '../../data_temp/Single-Genus/Pectobacterium/61.fna', '../../data_temp/Single-Genus/Pectobacterium/36.fna', '../../data_temp/Single-Genus/Pectobacterium/81.fna', '../../data_temp/Single-Genus/Pectobacterium/96.fna', '../../data_temp/Single-Genus/Pectobacterium/76.fna', '../../data_temp/Single-Genus/Pectobacterium/83.fna']\n",
      "Using previously computed OrthoANI values\n",
      "ITERATION 1 : kmer_length=10, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 2 : kmer_length=11, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 3 : kmer_length=12, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 4 : kmer_length=13, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 5 : kmer_length=14, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 6 : kmer_length=15, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 7 : kmer_length=16, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 8 : kmer_length=17, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 9 : kmer_length=18, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 10 : kmer_length=19, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 11 : kmer_length=20, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 12 : kmer_length=21, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 13 : kmer_length=22, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 14 : kmer_length=23, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 15 : kmer_length=24, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 16 : kmer_length=25, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 17 : kmer_length=26, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 18 : kmer_length=27, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 19 : kmer_length=28, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 20 : kmer_length=29, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 21 : kmer_length=30, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 22 : kmer_length=31, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 23 : kmer_length=32, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 24 : kmer_length=33, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 25 : kmer_length=34, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 26 : kmer_length=35, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "analysis_type='Single-Genus', analysis_data_dir='Morganella'\n",
      "Using genome files ['../../data_temp/Single-Genus/Morganella/64.fna', '../../data_temp/Single-Genus/Morganella/32.fna', '../../data_temp/Single-Genus/Morganella/99.fna', '../../data_temp/Single-Genus/Morganella/43.fna', '../../data_temp/Single-Genus/Morganella/7.fna', '../../data_temp/Single-Genus/Morganella/98.fna', '../../data_temp/Single-Genus/Morganella/58.fna', '../../data_temp/Single-Genus/Morganella/90.fna', '../../data_temp/Single-Genus/Morganella/14.fna', '../../data_temp/Single-Genus/Morganella/41.fna', '../../data_temp/Single-Genus/Morganella/67.fna', '../../data_temp/Single-Genus/Morganella/73.fna', '../../data_temp/Single-Genus/Morganella/17.fna', '../../data_temp/Single-Genus/Morganella/56.fna', '../../data_temp/Single-Genus/Morganella/50.fna', '../../data_temp/Single-Genus/Morganella/2.fna', '../../data_temp/Single-Genus/Morganella/46.fna', '../../data_temp/Single-Genus/Morganella/52.fna', '../../data_temp/Single-Genus/Morganella/65.fna', '../../data_temp/Single-Genus/Morganella/95.fna', '../../data_temp/Single-Genus/Morganella/45.fna', '../../data_temp/Single-Genus/Morganella/74.fna', '../../data_temp/Single-Genus/Morganella/44.fna', '../../data_temp/Single-Genus/Morganella/12.fna', '../../data_temp/Single-Genus/Morganella/94.fna', '../../data_temp/Single-Genus/Morganella/25.fna', '../../data_temp/Single-Genus/Morganella/60.fna', '../../data_temp/Single-Genus/Morganella/39.fna', '../../data_temp/Single-Genus/Morganella/22.fna', '../../data_temp/Single-Genus/Morganella/48.fna', '../../data_temp/Single-Genus/Morganella/88.fna', '../../data_temp/Single-Genus/Morganella/24.fna', '../../data_temp/Single-Genus/Morganella/9.fna', '../../data_temp/Single-Genus/Morganella/86.fna', '../../data_temp/Single-Genus/Morganella/5.fna', '../../data_temp/Single-Genus/Morganella/89.fna', '../../data_temp/Single-Genus/Morganella/71.fna', '../../data_temp/Single-Genus/Morganella/54.fna', '../../data_temp/Single-Genus/Morganella/23.fna', '../../data_temp/Single-Genus/Morganella/49.fna', '../../data_temp/Single-Genus/Morganella/40.fna', '../../data_temp/Single-Genus/Morganella/18.fna', '../../data_temp/Single-Genus/Morganella/4.fna', '../../data_temp/Single-Genus/Morganella/20.fna', '../../data_temp/Single-Genus/Morganella/63.fna', '../../data_temp/Single-Genus/Morganella/33.fna', '../../data_temp/Single-Genus/Morganella/87.fna', '../../data_temp/Single-Genus/Morganella/80.fna', '../../data_temp/Single-Genus/Morganella/26.fna', '../../data_temp/Single-Genus/Morganella/1.fna', '../../data_temp/Single-Genus/Morganella/47.fna', '../../data_temp/Single-Genus/Morganella/92.fna', '../../data_temp/Single-Genus/Morganella/51.fna', '../../data_temp/Single-Genus/Morganella/37.fna', '../../data_temp/Single-Genus/Morganella/84.fna', '../../data_temp/Single-Genus/Morganella/53.fna', '../../data_temp/Single-Genus/Morganella/70.fna', '../../data_temp/Single-Genus/Morganella/42.fna', '../../data_temp/Single-Genus/Morganella/57.fna', '../../data_temp/Single-Genus/Morganella/72.fna', '../../data_temp/Single-Genus/Morganella/91.fna', '../../data_temp/Single-Genus/Morganella/27.fna', '../../data_temp/Single-Genus/Morganella/19.fna', '../../data_temp/Single-Genus/Morganella/66.fna', '../../data_temp/Single-Genus/Morganella/13.fna', '../../data_temp/Single-Genus/Morganella/16.fna', '../../data_temp/Single-Genus/Morganella/29.fna', '../../data_temp/Single-Genus/Morganella/8.fna', '../../data_temp/Single-Genus/Morganella/21.fna', '../../data_temp/Single-Genus/Morganella/3.fna', '../../data_temp/Single-Genus/Morganella/10.fna', '../../data_temp/Single-Genus/Morganella/30.fna', '../../data_temp/Single-Genus/Morganella/38.fna', '../../data_temp/Single-Genus/Morganella/93.fna', '../../data_temp/Single-Genus/Morganella/78.fna', '../../data_temp/Single-Genus/Morganella/77.fna', '../../data_temp/Single-Genus/Morganella/75.fna', '../../data_temp/Single-Genus/Morganella/35.fna', '../../data_temp/Single-Genus/Morganella/31.fna', '../../data_temp/Single-Genus/Morganella/28.fna', '../../data_temp/Single-Genus/Morganella/82.fna', '../../data_temp/Single-Genus/Morganella/59.fna', '../../data_temp/Single-Genus/Morganella/69.fna', '../../data_temp/Single-Genus/Morganella/62.fna', '../../data_temp/Single-Genus/Morganella/85.fna', '../../data_temp/Single-Genus/Morganella/0.fna', '../../data_temp/Single-Genus/Morganella/79.fna', '../../data_temp/Single-Genus/Morganella/11.fna', '../../data_temp/Single-Genus/Morganella/34.fna', '../../data_temp/Single-Genus/Morganella/6.fna', '../../data_temp/Single-Genus/Morganella/68.fna', '../../data_temp/Single-Genus/Morganella/97.fna', '../../data_temp/Single-Genus/Morganella/15.fna', '../../data_temp/Single-Genus/Morganella/55.fna', '../../data_temp/Single-Genus/Morganella/61.fna', '../../data_temp/Single-Genus/Morganella/36.fna', '../../data_temp/Single-Genus/Morganella/81.fna', '../../data_temp/Single-Genus/Morganella/96.fna', '../../data_temp/Single-Genus/Morganella/76.fna', '../../data_temp/Single-Genus/Morganella/83.fna']\n",
      "Using previously computed OrthoANI values\n",
      "ITERATION 27 : kmer_length=10, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 28 : kmer_length=11, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 29 : kmer_length=12, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 30 : kmer_length=13, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 31 : kmer_length=14, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 32 : kmer_length=15, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 33 : kmer_length=16, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 34 : kmer_length=17, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 35 : kmer_length=18, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 36 : kmer_length=19, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 37 : kmer_length=20, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 38 : kmer_length=21, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 39 : kmer_length=22, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 40 : kmer_length=23, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 41 : kmer_length=24, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 42 : kmer_length=25, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 43 : kmer_length=26, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 44 : kmer_length=27, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 45 : kmer_length=28, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 46 : kmer_length=29, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 47 : kmer_length=30, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 48 : kmer_length=31, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 49 : kmer_length=32, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 50 : kmer_length=33, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 51 : kmer_length=34, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 52 : kmer_length=35, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "analysis_type='Single-Genus', analysis_data_dir='Xylella'\n",
      "Using genome files ['../../data_temp/Single-Genus/Xylella/64.fna', '../../data_temp/Single-Genus/Xylella/32.fna', '../../data_temp/Single-Genus/Xylella/99.fna', '../../data_temp/Single-Genus/Xylella/43.fna', '../../data_temp/Single-Genus/Xylella/7.fna', '../../data_temp/Single-Genus/Xylella/98.fna', '../../data_temp/Single-Genus/Xylella/58.fna', '../../data_temp/Single-Genus/Xylella/90.fna', '../../data_temp/Single-Genus/Xylella/14.fna', '../../data_temp/Single-Genus/Xylella/41.fna', '../../data_temp/Single-Genus/Xylella/67.fna', '../../data_temp/Single-Genus/Xylella/73.fna', '../../data_temp/Single-Genus/Xylella/17.fna', '../../data_temp/Single-Genus/Xylella/56.fna', '../../data_temp/Single-Genus/Xylella/50.fna', '../../data_temp/Single-Genus/Xylella/2.fna', '../../data_temp/Single-Genus/Xylella/46.fna', '../../data_temp/Single-Genus/Xylella/52.fna', '../../data_temp/Single-Genus/Xylella/65.fna', '../../data_temp/Single-Genus/Xylella/95.fna', '../../data_temp/Single-Genus/Xylella/45.fna', '../../data_temp/Single-Genus/Xylella/74.fna', '../../data_temp/Single-Genus/Xylella/44.fna', '../../data_temp/Single-Genus/Xylella/12.fna', '../../data_temp/Single-Genus/Xylella/94.fna', '../../data_temp/Single-Genus/Xylella/25.fna', '../../data_temp/Single-Genus/Xylella/60.fna', '../../data_temp/Single-Genus/Xylella/39.fna', '../../data_temp/Single-Genus/Xylella/22.fna', '../../data_temp/Single-Genus/Xylella/48.fna', '../../data_temp/Single-Genus/Xylella/88.fna', '../../data_temp/Single-Genus/Xylella/24.fna', '../../data_temp/Single-Genus/Xylella/9.fna', '../../data_temp/Single-Genus/Xylella/86.fna', '../../data_temp/Single-Genus/Xylella/5.fna', '../../data_temp/Single-Genus/Xylella/89.fna', '../../data_temp/Single-Genus/Xylella/71.fna', '../../data_temp/Single-Genus/Xylella/54.fna', '../../data_temp/Single-Genus/Xylella/23.fna', '../../data_temp/Single-Genus/Xylella/49.fna', '../../data_temp/Single-Genus/Xylella/40.fna', '../../data_temp/Single-Genus/Xylella/18.fna', '../../data_temp/Single-Genus/Xylella/4.fna', '../../data_temp/Single-Genus/Xylella/20.fna', '../../data_temp/Single-Genus/Xylella/63.fna', '../../data_temp/Single-Genus/Xylella/33.fna', '../../data_temp/Single-Genus/Xylella/87.fna', '../../data_temp/Single-Genus/Xylella/80.fna', '../../data_temp/Single-Genus/Xylella/26.fna', '../../data_temp/Single-Genus/Xylella/1.fna', '../../data_temp/Single-Genus/Xylella/47.fna', '../../data_temp/Single-Genus/Xylella/92.fna', '../../data_temp/Single-Genus/Xylella/51.fna', '../../data_temp/Single-Genus/Xylella/37.fna', '../../data_temp/Single-Genus/Xylella/84.fna', '../../data_temp/Single-Genus/Xylella/53.fna', '../../data_temp/Single-Genus/Xylella/70.fna', '../../data_temp/Single-Genus/Xylella/42.fna', '../../data_temp/Single-Genus/Xylella/57.fna', '../../data_temp/Single-Genus/Xylella/72.fna', '../../data_temp/Single-Genus/Xylella/91.fna', '../../data_temp/Single-Genus/Xylella/27.fna', '../../data_temp/Single-Genus/Xylella/19.fna', '../../data_temp/Single-Genus/Xylella/66.fna', '../../data_temp/Single-Genus/Xylella/13.fna', '../../data_temp/Single-Genus/Xylella/16.fna', '../../data_temp/Single-Genus/Xylella/29.fna', '../../data_temp/Single-Genus/Xylella/8.fna', '../../data_temp/Single-Genus/Xylella/21.fna', '../../data_temp/Single-Genus/Xylella/3.fna', '../../data_temp/Single-Genus/Xylella/10.fna', '../../data_temp/Single-Genus/Xylella/30.fna', '../../data_temp/Single-Genus/Xylella/38.fna', '../../data_temp/Single-Genus/Xylella/93.fna', '../../data_temp/Single-Genus/Xylella/78.fna', '../../data_temp/Single-Genus/Xylella/77.fna', '../../data_temp/Single-Genus/Xylella/75.fna', '../../data_temp/Single-Genus/Xylella/35.fna', '../../data_temp/Single-Genus/Xylella/31.fna', '../../data_temp/Single-Genus/Xylella/28.fna', '../../data_temp/Single-Genus/Xylella/82.fna', '../../data_temp/Single-Genus/Xylella/59.fna', '../../data_temp/Single-Genus/Xylella/69.fna', '../../data_temp/Single-Genus/Xylella/62.fna', '../../data_temp/Single-Genus/Xylella/85.fna', '../../data_temp/Single-Genus/Xylella/0.fna', '../../data_temp/Single-Genus/Xylella/79.fna', '../../data_temp/Single-Genus/Xylella/11.fna', '../../data_temp/Single-Genus/Xylella/34.fna', '../../data_temp/Single-Genus/Xylella/6.fna', '../../data_temp/Single-Genus/Xylella/68.fna', '../../data_temp/Single-Genus/Xylella/97.fna', '../../data_temp/Single-Genus/Xylella/15.fna', '../../data_temp/Single-Genus/Xylella/55.fna', '../../data_temp/Single-Genus/Xylella/61.fna', '../../data_temp/Single-Genus/Xylella/36.fna', '../../data_temp/Single-Genus/Xylella/81.fna', '../../data_temp/Single-Genus/Xylella/96.fna', '../../data_temp/Single-Genus/Xylella/76.fna', '../../data_temp/Single-Genus/Xylella/83.fna']\n",
      "Using previously computed OrthoANI values\n",
      "ITERATION 53 : kmer_length=10, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 54 : kmer_length=11, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 55 : kmer_length=12, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 56 : kmer_length=13, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 57 : kmer_length=14, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 58 : kmer_length=15, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 59 : kmer_length=16, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 60 : kmer_length=17, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 61 : kmer_length=18, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 62 : kmer_length=19, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 63 : kmer_length=20, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 64 : kmer_length=21, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 65 : kmer_length=22, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 66 : kmer_length=23, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 67 : kmer_length=24, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 68 : kmer_length=25, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 69 : kmer_length=26, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 70 : kmer_length=27, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 71 : kmer_length=28, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 72 : kmer_length=29, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 73 : kmer_length=30, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 74 : kmer_length=31, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 75 : kmer_length=32, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 76 : kmer_length=33, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 77 : kmer_length=34, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "ITERATION 78 : kmer_length=35, c_val=200, canon=True\n",
      "Using previously computed ANI values\n",
      "Computing correlations...\n",
      "analysis_type='Single-Family', analysis_data_dir='Enterobacteriaceae'\n",
      "Using genome files ['../../data_temp/Single-Family/Enterobacteriaceae/479.fna', '../../data_temp/Single-Family/Enterobacteriaceae/306.fna', '../../data_temp/Single-Family/Enterobacteriaceae/228.fna', '../../data_temp/Single-Family/Enterobacteriaceae/404.fna', '../../data_temp/Single-Family/Enterobacteriaceae/112.fna', '../../data_temp/Single-Family/Enterobacteriaceae/164.fna', '../../data_temp/Single-Family/Enterobacteriaceae/429.fna', '../../data_temp/Single-Family/Enterobacteriaceae/367.fna', '../../data_temp/Single-Family/Enterobacteriaceae/64.fna', '../../data_temp/Single-Family/Enterobacteriaceae/491.fna', '../../data_temp/Single-Family/Enterobacteriaceae/236.fna', '../../data_temp/Single-Family/Enterobacteriaceae/130.fna', '../../data_temp/Single-Family/Enterobacteriaceae/281.fna', '../../data_temp/Single-Family/Enterobacteriaceae/287.fna', '../../data_temp/Single-Family/Enterobacteriaceae/413.fna', '../../data_temp/Single-Family/Enterobacteriaceae/300.fna', '../../data_temp/Single-Family/Enterobacteriaceae/252.fna', '../../data_temp/Single-Family/Enterobacteriaceae/140.fna', '../../data_temp/Single-Family/Enterobacteriaceae/32.fna', '../../data_temp/Single-Family/Enterobacteriaceae/379.fna', '../../data_temp/Single-Family/Enterobacteriaceae/222.fna', '../../data_temp/Single-Family/Enterobacteriaceae/153.fna', '../../data_temp/Single-Family/Enterobacteriaceae/350.fna', '../../data_temp/Single-Family/Enterobacteriaceae/102.fna', '../../data_temp/Single-Family/Enterobacteriaceae/430.fna', '../../data_temp/Single-Family/Enterobacteriaceae/488.fna', '../../data_temp/Single-Family/Enterobacteriaceae/99.fna', '../../data_temp/Single-Family/Enterobacteriaceae/172.fna', '../../data_temp/Single-Family/Enterobacteriaceae/446.fna', '../../data_temp/Single-Family/Enterobacteriaceae/448.fna', '../../data_temp/Single-Family/Enterobacteriaceae/385.fna', '../../data_temp/Single-Family/Enterobacteriaceae/152.fna', '../../data_temp/Single-Family/Enterobacteriaceae/316.fna', '../../data_temp/Single-Family/Enterobacteriaceae/162.fna', '../../data_temp/Single-Family/Enterobacteriaceae/321.fna', '../../data_temp/Single-Family/Enterobacteriaceae/439.fna', '../../data_temp/Single-Family/Enterobacteriaceae/128.fna', '../../data_temp/Single-Family/Enterobacteriaceae/314.fna', '../../data_temp/Single-Family/Enterobacteriaceae/191.fna', '../../data_temp/Single-Family/Enterobacteriaceae/177.fna', '../../data_temp/Single-Family/Enterobacteriaceae/43.fna', '../../data_temp/Single-Family/Enterobacteriaceae/212.fna', '../../data_temp/Single-Family/Enterobacteriaceae/260.fna', '../../data_temp/Single-Family/Enterobacteriaceae/452.fna', '../../data_temp/Single-Family/Enterobacteriaceae/315.fna', '../../data_temp/Single-Family/Enterobacteriaceae/421.fna', '../../data_temp/Single-Family/Enterobacteriaceae/384.fna', '../../data_temp/Single-Family/Enterobacteriaceae/7.fna', '../../data_temp/Single-Family/Enterobacteriaceae/98.fna', '../../data_temp/Single-Family/Enterobacteriaceae/213.fna', '../../data_temp/Single-Family/Enterobacteriaceae/58.fna', '../../data_temp/Single-Family/Enterobacteriaceae/425.fna', '../../data_temp/Single-Family/Enterobacteriaceae/392.fna', '../../data_temp/Single-Family/Enterobacteriaceae/493.fna', '../../data_temp/Single-Family/Enterobacteriaceae/283.fna', '../../data_temp/Single-Family/Enterobacteriaceae/90.fna', '../../data_temp/Single-Family/Enterobacteriaceae/473.fna', '../../data_temp/Single-Family/Enterobacteriaceae/440.fna', '../../data_temp/Single-Family/Enterobacteriaceae/147.fna', '../../data_temp/Single-Family/Enterobacteriaceae/133.fna', '../../data_temp/Single-Family/Enterobacteriaceae/168.fna', '../../data_temp/Single-Family/Enterobacteriaceae/266.fna', '../../data_temp/Single-Family/Enterobacteriaceae/149.fna', '../../data_temp/Single-Family/Enterobacteriaceae/221.fna', '../../data_temp/Single-Family/Enterobacteriaceae/151.fna', '../../data_temp/Single-Family/Enterobacteriaceae/141.fna', '../../data_temp/Single-Family/Enterobacteriaceae/347.fna', '../../data_temp/Single-Family/Enterobacteriaceae/135.fna', '../../data_temp/Single-Family/Enterobacteriaceae/233.fna', '../../data_temp/Single-Family/Enterobacteriaceae/492.fna', '../../data_temp/Single-Family/Enterobacteriaceae/14.fna', '../../data_temp/Single-Family/Enterobacteriaceae/363.fna', '../../data_temp/Single-Family/Enterobacteriaceae/499.fna', '../../data_temp/Single-Family/Enterobacteriaceae/41.fna', '../../data_temp/Single-Family/Enterobacteriaceae/424.fna', '../../data_temp/Single-Family/Enterobacteriaceae/178.fna', '../../data_temp/Single-Family/Enterobacteriaceae/157.fna', '../../data_temp/Single-Family/Enterobacteriaceae/254.fna', '../../data_temp/Single-Family/Enterobacteriaceae/214.fna', '../../data_temp/Single-Family/Enterobacteriaceae/67.fna', '../../data_temp/Single-Family/Enterobacteriaceae/105.fna', '../../data_temp/Single-Family/Enterobacteriaceae/100.fna', '../../data_temp/Single-Family/Enterobacteriaceae/73.fna', '../../data_temp/Single-Family/Enterobacteriaceae/119.fna', '../../data_temp/Single-Family/Enterobacteriaceae/17.fna', '../../data_temp/Single-Family/Enterobacteriaceae/209.fna', '../../data_temp/Single-Family/Enterobacteriaceae/332.fna', '../../data_temp/Single-Family/Enterobacteriaceae/56.fna', '../../data_temp/Single-Family/Enterobacteriaceae/308.fna', '../../data_temp/Single-Family/Enterobacteriaceae/390.fna', '../../data_temp/Single-Family/Enterobacteriaceae/50.fna', '../../data_temp/Single-Family/Enterobacteriaceae/428.fna', '../../data_temp/Single-Family/Enterobacteriaceae/2.fna', '../../data_temp/Single-Family/Enterobacteriaceae/388.fna', '../../data_temp/Single-Family/Enterobacteriaceae/46.fna', '../../data_temp/Single-Family/Enterobacteriaceae/349.fna', '../../data_temp/Single-Family/Enterobacteriaceae/242.fna', '../../data_temp/Single-Family/Enterobacteriaceae/237.fna', '../../data_temp/Single-Family/Enterobacteriaceae/52.fna', '../../data_temp/Single-Family/Enterobacteriaceae/270.fna', '../../data_temp/Single-Family/Enterobacteriaceae/155.fna', '../../data_temp/Single-Family/Enterobacteriaceae/344.fna', '../../data_temp/Single-Family/Enterobacteriaceae/65.fna', '../../data_temp/Single-Family/Enterobacteriaceae/123.fna', '../../data_temp/Single-Family/Enterobacteriaceae/357.fna', '../../data_temp/Single-Family/Enterobacteriaceae/487.fna', '../../data_temp/Single-Family/Enterobacteriaceae/113.fna', '../../data_temp/Single-Family/Enterobacteriaceae/362.fna', '../../data_temp/Single-Family/Enterobacteriaceae/378.fna', '../../data_temp/Single-Family/Enterobacteriaceae/313.fna', '../../data_temp/Single-Family/Enterobacteriaceae/394.fna', '../../data_temp/Single-Family/Enterobacteriaceae/364.fna', '../../data_temp/Single-Family/Enterobacteriaceae/481.fna', '../../data_temp/Single-Family/Enterobacteriaceae/497.fna', '../../data_temp/Single-Family/Enterobacteriaceae/386.fna', '../../data_temp/Single-Family/Enterobacteriaceae/211.fna', '../../data_temp/Single-Family/Enterobacteriaceae/181.fna', '../../data_temp/Single-Family/Enterobacteriaceae/95.fna', '../../data_temp/Single-Family/Enterobacteriaceae/45.fna', '../../data_temp/Single-Family/Enterobacteriaceae/249.fna', '../../data_temp/Single-Family/Enterobacteriaceae/120.fna', '../../data_temp/Single-Family/Enterobacteriaceae/495.fna', '../../data_temp/Single-Family/Enterobacteriaceae/319.fna', '../../data_temp/Single-Family/Enterobacteriaceae/125.fna', '../../data_temp/Single-Family/Enterobacteriaceae/189.fna', '../../data_temp/Single-Family/Enterobacteriaceae/171.fna', '../../data_temp/Single-Family/Enterobacteriaceae/74.fna', '../../data_temp/Single-Family/Enterobacteriaceae/335.fna', '../../data_temp/Single-Family/Enterobacteriaceae/253.fna', '../../data_temp/Single-Family/Enterobacteriaceae/467.fna', '../../data_temp/Single-Family/Enterobacteriaceae/44.fna', '../../data_temp/Single-Family/Enterobacteriaceae/342.fna', '../../data_temp/Single-Family/Enterobacteriaceae/431.fna', '../../data_temp/Single-Family/Enterobacteriaceae/159.fna', '../../data_temp/Single-Family/Enterobacteriaceae/498.fna', '../../data_temp/Single-Family/Enterobacteriaceae/12.fna', '../../data_temp/Single-Family/Enterobacteriaceae/450.fna', '../../data_temp/Single-Family/Enterobacteriaceae/154.fna', '../../data_temp/Single-Family/Enterobacteriaceae/445.fna', '../../data_temp/Single-Family/Enterobacteriaceae/200.fna', '../../data_temp/Single-Family/Enterobacteriaceae/496.fna', '../../data_temp/Single-Family/Enterobacteriaceae/343.fna', '../../data_temp/Single-Family/Enterobacteriaceae/276.fna', '../../data_temp/Single-Family/Enterobacteriaceae/365.fna', '../../data_temp/Single-Family/Enterobacteriaceae/148.fna', '../../data_temp/Single-Family/Enterobacteriaceae/94.fna', '../../data_temp/Single-Family/Enterobacteriaceae/25.fna', '../../data_temp/Single-Family/Enterobacteriaceae/375.fna', '../../data_temp/Single-Family/Enterobacteriaceae/460.fna', '../../data_temp/Single-Family/Enterobacteriaceae/188.fna', '../../data_temp/Single-Family/Enterobacteriaceae/475.fna', '../../data_temp/Single-Family/Enterobacteriaceae/320.fna', '../../data_temp/Single-Family/Enterobacteriaceae/437.fna', '../../data_temp/Single-Family/Enterobacteriaceae/195.fna', '../../data_temp/Single-Family/Enterobacteriaceae/382.fna', '../../data_temp/Single-Family/Enterobacteriaceae/480.fna', '../../data_temp/Single-Family/Enterobacteriaceae/124.fna', '../../data_temp/Single-Family/Enterobacteriaceae/216.fna', '../../data_temp/Single-Family/Enterobacteriaceae/60.fna', '../../data_temp/Single-Family/Enterobacteriaceae/39.fna', '../../data_temp/Single-Family/Enterobacteriaceae/22.fna', '../../data_temp/Single-Family/Enterobacteriaceae/302.fna', '../../data_temp/Single-Family/Enterobacteriaceae/261.fna', '../../data_temp/Single-Family/Enterobacteriaceae/108.fna', '../../data_temp/Single-Family/Enterobacteriaceae/248.fna', '../../data_temp/Single-Family/Enterobacteriaceae/393.fna', '../../data_temp/Single-Family/Enterobacteriaceae/223.fna', '../../data_temp/Single-Family/Enterobacteriaceae/230.fna', '../../data_temp/Single-Family/Enterobacteriaceae/48.fna', '../../data_temp/Single-Family/Enterobacteriaceae/88.fna', '../../data_temp/Single-Family/Enterobacteriaceae/328.fna', '../../data_temp/Single-Family/Enterobacteriaceae/24.fna', '../../data_temp/Single-Family/Enterobacteriaceae/418.fna', '../../data_temp/Single-Family/Enterobacteriaceae/203.fna', '../../data_temp/Single-Family/Enterobacteriaceae/298.fna', '../../data_temp/Single-Family/Enterobacteriaceae/438.fna', '../../data_temp/Single-Family/Enterobacteriaceae/436.fna', '../../data_temp/Single-Family/Enterobacteriaceae/336.fna', '../../data_temp/Single-Family/Enterobacteriaceae/156.fna', '../../data_temp/Single-Family/Enterobacteriaceae/9.fna', '../../data_temp/Single-Family/Enterobacteriaceae/354.fna', '../../data_temp/Single-Family/Enterobacteriaceae/184.fna', '../../data_temp/Single-Family/Enterobacteriaceae/226.fna', '../../data_temp/Single-Family/Enterobacteriaceae/294.fna', '../../data_temp/Single-Family/Enterobacteriaceae/194.fna', '../../data_temp/Single-Family/Enterobacteriaceae/161.fna', '../../data_temp/Single-Family/Enterobacteriaceae/234.fna', '../../data_temp/Single-Family/Enterobacteriaceae/482.fna', '../../data_temp/Single-Family/Enterobacteriaceae/86.fna', '../../data_temp/Single-Family/Enterobacteriaceae/5.fna', '../../data_temp/Single-Family/Enterobacteriaceae/433.fna', '../../data_temp/Single-Family/Enterobacteriaceae/304.fna', '../../data_temp/Single-Family/Enterobacteriaceae/317.fna', '../../data_temp/Single-Family/Enterobacteriaceae/417.fna', '../../data_temp/Single-Family/Enterobacteriaceae/464.fna', '../../data_temp/Single-Family/Enterobacteriaceae/339.fna', '../../data_temp/Single-Family/Enterobacteriaceae/301.fna', '../../data_temp/Single-Family/Enterobacteriaceae/396.fna', '../../data_temp/Single-Family/Enterobacteriaceae/225.fna', '../../data_temp/Single-Family/Enterobacteriaceae/175.fna', '../../data_temp/Single-Family/Enterobacteriaceae/89.fna', '../../data_temp/Single-Family/Enterobacteriaceae/255.fna', '../../data_temp/Single-Family/Enterobacteriaceae/190.fna', '../../data_temp/Single-Family/Enterobacteriaceae/399.fna', '../../data_temp/Single-Family/Enterobacteriaceae/456.fna', '../../data_temp/Single-Family/Enterobacteriaceae/274.fna', '../../data_temp/Single-Family/Enterobacteriaceae/282.fna', '../../data_temp/Single-Family/Enterobacteriaceae/400.fna', '../../data_temp/Single-Family/Enterobacteriaceae/71.fna', '../../data_temp/Single-Family/Enterobacteriaceae/158.fna', '../../data_temp/Single-Family/Enterobacteriaceae/416.fna', '../../data_temp/Single-Family/Enterobacteriaceae/405.fna', '../../data_temp/Single-Family/Enterobacteriaceae/387.fna', '../../data_temp/Single-Family/Enterobacteriaceae/494.fna', '../../data_temp/Single-Family/Enterobacteriaceae/192.fna', '../../data_temp/Single-Family/Enterobacteriaceae/380.fna', '../../data_temp/Single-Family/Enterobacteriaceae/186.fna', '../../data_temp/Single-Family/Enterobacteriaceae/264.fna', '../../data_temp/Single-Family/Enterobacteriaceae/54.fna', '../../data_temp/Single-Family/Enterobacteriaceae/284.fna', '../../data_temp/Single-Family/Enterobacteriaceae/325.fna', '../../data_temp/Single-Family/Enterobacteriaceae/23.fna', '../../data_temp/Single-Family/Enterobacteriaceae/408.fna', '../../data_temp/Single-Family/Enterobacteriaceae/49.fna', '../../data_temp/Single-Family/Enterobacteriaceae/303.fna', '../../data_temp/Single-Family/Enterobacteriaceae/121.fna', '../../data_temp/Single-Family/Enterobacteriaceae/389.fna', '../../data_temp/Single-Family/Enterobacteriaceae/246.fna', '../../data_temp/Single-Family/Enterobacteriaceae/251.fna', '../../data_temp/Single-Family/Enterobacteriaceae/420.fna', '../../data_temp/Single-Family/Enterobacteriaceae/455.fna', '../../data_temp/Single-Family/Enterobacteriaceae/346.fna', '../../data_temp/Single-Family/Enterobacteriaceae/40.fna', '../../data_temp/Single-Family/Enterobacteriaceae/205.fna', '../../data_temp/Single-Family/Enterobacteriaceae/201.fna', '../../data_temp/Single-Family/Enterobacteriaceae/18.fna', '../../data_temp/Single-Family/Enterobacteriaceae/307.fna', '../../data_temp/Single-Family/Enterobacteriaceae/257.fna', '../../data_temp/Single-Family/Enterobacteriaceae/170.fna', '../../data_temp/Single-Family/Enterobacteriaceae/4.fna', '../../data_temp/Single-Family/Enterobacteriaceae/210.fna', '../../data_temp/Single-Family/Enterobacteriaceae/206.fna', '../../data_temp/Single-Family/Enterobacteriaceae/143.fna', '../../data_temp/Single-Family/Enterobacteriaceae/20.fna', '../../data_temp/Single-Family/Enterobacteriaceae/63.fna', '../../data_temp/Single-Family/Enterobacteriaceae/355.fna', '../../data_temp/Single-Family/Enterobacteriaceae/414.fna', '../../data_temp/Single-Family/Enterobacteriaceae/33.fna', '../../data_temp/Single-Family/Enterobacteriaceae/163.fna', '../../data_temp/Single-Family/Enterobacteriaceae/198.fna', '../../data_temp/Single-Family/Enterobacteriaceae/87.fna', '../../data_temp/Single-Family/Enterobacteriaceae/377.fna', '../../data_temp/Single-Family/Enterobacteriaceae/132.fna', '../../data_temp/Single-Family/Enterobacteriaceae/80.fna', '../../data_temp/Single-Family/Enterobacteriaceae/26.fna', '../../data_temp/Single-Family/Enterobacteriaceae/279.fna', '../../data_temp/Single-Family/Enterobacteriaceae/402.fna', '../../data_temp/Single-Family/Enterobacteriaceae/150.fna', '../../data_temp/Single-Family/Enterobacteriaceae/1.fna', '../../data_temp/Single-Family/Enterobacteriaceae/115.fna', '../../data_temp/Single-Family/Enterobacteriaceae/333.fna', '../../data_temp/Single-Family/Enterobacteriaceae/410.fna', '../../data_temp/Single-Family/Enterobacteriaceae/47.fna', '../../data_temp/Single-Family/Enterobacteriaceae/401.fna', '../../data_temp/Single-Family/Enterobacteriaceae/92.fna', '../../data_temp/Single-Family/Enterobacteriaceae/104.fna', '../../data_temp/Single-Family/Enterobacteriaceae/318.fna', '../../data_temp/Single-Family/Enterobacteriaceae/462.fna', '../../data_temp/Single-Family/Enterobacteriaceae/247.fna', '../../data_temp/Single-Family/Enterobacteriaceae/444.fna', '../../data_temp/Single-Family/Enterobacteriaceae/465.fna', '../../data_temp/Single-Family/Enterobacteriaceae/406.fna', '../../data_temp/Single-Family/Enterobacteriaceae/51.fna', '../../data_temp/Single-Family/Enterobacteriaceae/391.fna', '../../data_temp/Single-Family/Enterobacteriaceae/277.fna', '../../data_temp/Single-Family/Enterobacteriaceae/107.fna', '../../data_temp/Single-Family/Enterobacteriaceae/454.fna', '../../data_temp/Single-Family/Enterobacteriaceae/273.fna', '../../data_temp/Single-Family/Enterobacteriaceae/131.fna', '../../data_temp/Single-Family/Enterobacteriaceae/453.fna', '../../data_temp/Single-Family/Enterobacteriaceae/185.fna', '../../data_temp/Single-Family/Enterobacteriaceae/403.fna', '../../data_temp/Single-Family/Enterobacteriaceae/297.fna', '../../data_temp/Single-Family/Enterobacteriaceae/114.fna', '../../data_temp/Single-Family/Enterobacteriaceae/427.fna', '../../data_temp/Single-Family/Enterobacteriaceae/37.fna', '../../data_temp/Single-Family/Enterobacteriaceae/84.fna', '../../data_temp/Single-Family/Enterobacteriaceae/471.fna', '../../data_temp/Single-Family/Enterobacteriaceae/53.fna', '../../data_temp/Single-Family/Enterobacteriaceae/468.fna', '../../data_temp/Single-Family/Enterobacteriaceae/240.fna', '../../data_temp/Single-Family/Enterobacteriaceae/285.fna', '../../data_temp/Single-Family/Enterobacteriaceae/207.fna', '../../data_temp/Single-Family/Enterobacteriaceae/169.fna', '../../data_temp/Single-Family/Enterobacteriaceae/70.fna', '../../data_temp/Single-Family/Enterobacteriaceae/199.fna', '../../data_temp/Single-Family/Enterobacteriaceae/111.fna', '../../data_temp/Single-Family/Enterobacteriaceae/474.fna', '../../data_temp/Single-Family/Enterobacteriaceae/483.fna', '../../data_temp/Single-Family/Enterobacteriaceae/42.fna', '../../data_temp/Single-Family/Enterobacteriaceae/244.fna', '../../data_temp/Single-Family/Enterobacteriaceae/117.fna', '../../data_temp/Single-Family/Enterobacteriaceae/356.fna', '../../data_temp/Single-Family/Enterobacteriaceae/196.fna', '../../data_temp/Single-Family/Enterobacteriaceae/183.fna', '../../data_temp/Single-Family/Enterobacteriaceae/134.fna', '../../data_temp/Single-Family/Enterobacteriaceae/238.fna', '../../data_temp/Single-Family/Enterobacteriaceae/485.fna', '../../data_temp/Single-Family/Enterobacteriaceae/57.fna', '../../data_temp/Single-Family/Enterobacteriaceae/72.fna', '../../data_temp/Single-Family/Enterobacteriaceae/262.fna', '../../data_temp/Single-Family/Enterobacteriaceae/272.fna', '../../data_temp/Single-Family/Enterobacteriaceae/258.fna', '../../data_temp/Single-Family/Enterobacteriaceae/91.fna', '../../data_temp/Single-Family/Enterobacteriaceae/27.fna', '../../data_temp/Single-Family/Enterobacteriaceae/419.fna', '../../data_temp/Single-Family/Enterobacteriaceae/447.fna', '../../data_temp/Single-Family/Enterobacteriaceae/204.fna', '../../data_temp/Single-Family/Enterobacteriaceae/324.fna', '../../data_temp/Single-Family/Enterobacteriaceae/145.fna', '../../data_temp/Single-Family/Enterobacteriaceae/19.fna', '../../data_temp/Single-Family/Enterobacteriaceae/109.fna', '../../data_temp/Single-Family/Enterobacteriaceae/66.fna', '../../data_temp/Single-Family/Enterobacteriaceae/461.fna', '../../data_temp/Single-Family/Enterobacteriaceae/13.fna', '../../data_temp/Single-Family/Enterobacteriaceae/16.fna', '../../data_temp/Single-Family/Enterobacteriaceae/353.fna', '../../data_temp/Single-Family/Enterobacteriaceae/243.fna', '../../data_temp/Single-Family/Enterobacteriaceae/372.fna', '../../data_temp/Single-Family/Enterobacteriaceae/160.fna', '../../data_temp/Single-Family/Enterobacteriaceae/458.fna', '../../data_temp/Single-Family/Enterobacteriaceae/398.fna', '../../data_temp/Single-Family/Enterobacteriaceae/208.fna', '../../data_temp/Single-Family/Enterobacteriaceae/29.fna', '../../data_temp/Single-Family/Enterobacteriaceae/361.fna', '../../data_temp/Single-Family/Enterobacteriaceae/263.fna', '../../data_temp/Single-Family/Enterobacteriaceae/290.fna', '../../data_temp/Single-Family/Enterobacteriaceae/484.fna', '../../data_temp/Single-Family/Enterobacteriaceae/166.fna', '../../data_temp/Single-Family/Enterobacteriaceae/8.fna', '../../data_temp/Single-Family/Enterobacteriaceae/415.fna', '../../data_temp/Single-Family/Enterobacteriaceae/21.fna', '../../data_temp/Single-Family/Enterobacteriaceae/3.fna', '../../data_temp/Single-Family/Enterobacteriaceae/348.fna', '../../data_temp/Single-Family/Enterobacteriaceae/459.fna', '../../data_temp/Single-Family/Enterobacteriaceae/334.fna', '../../data_temp/Single-Family/Enterobacteriaceae/383.fna', '../../data_temp/Single-Family/Enterobacteriaceae/10.fna', '../../data_temp/Single-Family/Enterobacteriaceae/432.fna', '../../data_temp/Single-Family/Enterobacteriaceae/351.fna', '../../data_temp/Single-Family/Enterobacteriaceae/30.fna', '../../data_temp/Single-Family/Enterobacteriaceae/327.fna', '../../data_temp/Single-Family/Enterobacteriaceae/305.fna', '../../data_temp/Single-Family/Enterobacteriaceae/368.fna', '../../data_temp/Single-Family/Enterobacteriaceae/329.fna', '../../data_temp/Single-Family/Enterobacteriaceae/338.fna', '../../data_temp/Single-Family/Enterobacteriaceae/229.fna', '../../data_temp/Single-Family/Enterobacteriaceae/126.fna', '../../data_temp/Single-Family/Enterobacteriaceae/286.fna', '../../data_temp/Single-Family/Enterobacteriaceae/422.fna', '../../data_temp/Single-Family/Enterobacteriaceae/38.fna', '../../data_temp/Single-Family/Enterobacteriaceae/397.fna', '../../data_temp/Single-Family/Enterobacteriaceae/341.fna', '../../data_temp/Single-Family/Enterobacteriaceae/182.fna', '../../data_temp/Single-Family/Enterobacteriaceae/103.fna', '../../data_temp/Single-Family/Enterobacteriaceae/217.fna', '../../data_temp/Single-Family/Enterobacteriaceae/477.fna', '../../data_temp/Single-Family/Enterobacteriaceae/486.fna', '../../data_temp/Single-Family/Enterobacteriaceae/271.fna', '../../data_temp/Single-Family/Enterobacteriaceae/137.fna', '../../data_temp/Single-Family/Enterobacteriaceae/337.fna', '../../data_temp/Single-Family/Enterobacteriaceae/293.fna', '../../data_temp/Single-Family/Enterobacteriaceae/93.fna', '../../data_temp/Single-Family/Enterobacteriaceae/78.fna', '../../data_temp/Single-Family/Enterobacteriaceae/77.fna', '../../data_temp/Single-Family/Enterobacteriaceae/75.fna', '../../data_temp/Single-Family/Enterobacteriaceae/35.fna', '../../data_temp/Single-Family/Enterobacteriaceae/31.fna', '../../data_temp/Single-Family/Enterobacteriaceae/28.fna', '../../data_temp/Single-Family/Enterobacteriaceae/407.fna', '../../data_temp/Single-Family/Enterobacteriaceae/165.fna', '../../data_temp/Single-Family/Enterobacteriaceae/82.fna', '../../data_temp/Single-Family/Enterobacteriaceae/59.fna', '../../data_temp/Single-Family/Enterobacteriaceae/256.fna', '../../data_temp/Single-Family/Enterobacteriaceae/219.fna', '../../data_temp/Single-Family/Enterobacteriaceae/179.fna', '../../data_temp/Single-Family/Enterobacteriaceae/472.fna', '../../data_temp/Single-Family/Enterobacteriaceae/376.fna', '../../data_temp/Single-Family/Enterobacteriaceae/326.fna', '../../data_temp/Single-Family/Enterobacteriaceae/269.fna', '../../data_temp/Single-Family/Enterobacteriaceae/197.fna', '../../data_temp/Single-Family/Enterobacteriaceae/122.fna', '../../data_temp/Single-Family/Enterobacteriaceae/69.fna', '../../data_temp/Single-Family/Enterobacteriaceae/322.fna', '../../data_temp/Single-Family/Enterobacteriaceae/202.fna', '../../data_temp/Single-Family/Enterobacteriaceae/62.fna', '../../data_temp/Single-Family/Enterobacteriaceae/85.fna', '../../data_temp/Single-Family/Enterobacteriaceae/381.fna', '../../data_temp/Single-Family/Enterobacteriaceae/224.fna', '../../data_temp/Single-Family/Enterobacteriaceae/434.fna', '../../data_temp/Single-Family/Enterobacteriaceae/0.fna', '../../data_temp/Single-Family/Enterobacteriaceae/239.fna', '../../data_temp/Single-Family/Enterobacteriaceae/167.fna', '../../data_temp/Single-Family/Enterobacteriaceae/180.fna', '../../data_temp/Single-Family/Enterobacteriaceae/443.fna', '../../data_temp/Single-Family/Enterobacteriaceae/144.fna', '../../data_temp/Single-Family/Enterobacteriaceae/374.fna', '../../data_temp/Single-Family/Enterobacteriaceae/187.fna', '../../data_temp/Single-Family/Enterobacteriaceae/466.fna', '../../data_temp/Single-Family/Enterobacteriaceae/267.fna', '../../data_temp/Single-Family/Enterobacteriaceae/139.fna', '../../data_temp/Single-Family/Enterobacteriaceae/366.fna', '../../data_temp/Single-Family/Enterobacteriaceae/136.fna', '../../data_temp/Single-Family/Enterobacteriaceae/323.fna', '../../data_temp/Single-Family/Enterobacteriaceae/235.fna', '../../data_temp/Single-Family/Enterobacteriaceae/173.fna', '../../data_temp/Single-Family/Enterobacteriaceae/79.fna', '../../data_temp/Single-Family/Enterobacteriaceae/227.fna', '../../data_temp/Single-Family/Enterobacteriaceae/288.fna', '../../data_temp/Single-Family/Enterobacteriaceae/259.fna', '../../data_temp/Single-Family/Enterobacteriaceae/442.fna', '../../data_temp/Single-Family/Enterobacteriaceae/11.fna', '../../data_temp/Single-Family/Enterobacteriaceae/220.fna', '../../data_temp/Single-Family/Enterobacteriaceae/174.fna', '../../data_temp/Single-Family/Enterobacteriaceae/409.fna', '../../data_temp/Single-Family/Enterobacteriaceae/469.fna', '../../data_temp/Single-Family/Enterobacteriaceae/34.fna', '../../data_temp/Single-Family/Enterobacteriaceae/412.fna', '../../data_temp/Single-Family/Enterobacteriaceae/435.fna', '../../data_temp/Single-Family/Enterobacteriaceae/142.fna', '../../data_temp/Single-Family/Enterobacteriaceae/6.fna', '../../data_temp/Single-Family/Enterobacteriaceae/449.fna', '../../data_temp/Single-Family/Enterobacteriaceae/218.fna', '../../data_temp/Single-Family/Enterobacteriaceae/463.fna', '../../data_temp/Single-Family/Enterobacteriaceae/295.fna', '../../data_temp/Single-Family/Enterobacteriaceae/331.fna', '../../data_temp/Single-Family/Enterobacteriaceae/358.fna', '../../data_temp/Single-Family/Enterobacteriaceae/340.fna', '../../data_temp/Single-Family/Enterobacteriaceae/275.fna', '../../data_temp/Single-Family/Enterobacteriaceae/232.fna', '../../data_temp/Single-Family/Enterobacteriaceae/193.fna', '../../data_temp/Single-Family/Enterobacteriaceae/423.fna', '../../data_temp/Single-Family/Enterobacteriaceae/426.fna', '../../data_temp/Single-Family/Enterobacteriaceae/478.fna', '../../data_temp/Single-Family/Enterobacteriaceae/296.fna', '../../data_temp/Single-Family/Enterobacteriaceae/68.fna', '../../data_temp/Single-Family/Enterobacteriaceae/395.fna', '../../data_temp/Single-Family/Enterobacteriaceae/106.fna', '../../data_temp/Single-Family/Enterobacteriaceae/241.fna', '../../data_temp/Single-Family/Enterobacteriaceae/359.fna', '../../data_temp/Single-Family/Enterobacteriaceae/278.fna', '../../data_temp/Single-Family/Enterobacteriaceae/309.fna', '../../data_temp/Single-Family/Enterobacteriaceae/489.fna', '../../data_temp/Single-Family/Enterobacteriaceae/101.fna', '../../data_temp/Single-Family/Enterobacteriaceae/215.fna', '../../data_temp/Single-Family/Enterobacteriaceae/245.fna', '../../data_temp/Single-Family/Enterobacteriaceae/369.fna', '../../data_temp/Single-Family/Enterobacteriaceae/97.fna', '../../data_temp/Single-Family/Enterobacteriaceae/15.fna', '../../data_temp/Single-Family/Enterobacteriaceae/250.fna', '../../data_temp/Single-Family/Enterobacteriaceae/330.fna', '../../data_temp/Single-Family/Enterobacteriaceae/360.fna', '../../data_temp/Single-Family/Enterobacteriaceae/127.fna', '../../data_temp/Single-Family/Enterobacteriaceae/289.fna', '../../data_temp/Single-Family/Enterobacteriaceae/371.fna', '../../data_temp/Single-Family/Enterobacteriaceae/470.fna', '../../data_temp/Single-Family/Enterobacteriaceae/55.fna', '../../data_temp/Single-Family/Enterobacteriaceae/61.fna', '../../data_temp/Single-Family/Enterobacteriaceae/116.fna', '../../data_temp/Single-Family/Enterobacteriaceae/370.fna', '../../data_temp/Single-Family/Enterobacteriaceae/280.fna', '../../data_temp/Single-Family/Enterobacteriaceae/345.fna', '../../data_temp/Single-Family/Enterobacteriaceae/490.fna', '../../data_temp/Single-Family/Enterobacteriaceae/36.fna', '../../data_temp/Single-Family/Enterobacteriaceae/138.fna', '../../data_temp/Single-Family/Enterobacteriaceae/81.fna', '../../data_temp/Single-Family/Enterobacteriaceae/411.fna', '../../data_temp/Single-Family/Enterobacteriaceae/310.fna', '../../data_temp/Single-Family/Enterobacteriaceae/96.fna', '../../data_temp/Single-Family/Enterobacteriaceae/118.fna', '../../data_temp/Single-Family/Enterobacteriaceae/311.fna', '../../data_temp/Single-Family/Enterobacteriaceae/457.fna', '../../data_temp/Single-Family/Enterobacteriaceae/451.fna', '../../data_temp/Single-Family/Enterobacteriaceae/176.fna', '../../data_temp/Single-Family/Enterobacteriaceae/146.fna', '../../data_temp/Single-Family/Enterobacteriaceae/265.fna', '../../data_temp/Single-Family/Enterobacteriaceae/268.fna', '../../data_temp/Single-Family/Enterobacteriaceae/231.fna', '../../data_temp/Single-Family/Enterobacteriaceae/312.fna', '../../data_temp/Single-Family/Enterobacteriaceae/373.fna', '../../data_temp/Single-Family/Enterobacteriaceae/299.fna', '../../data_temp/Single-Family/Enterobacteriaceae/476.fna', '../../data_temp/Single-Family/Enterobacteriaceae/352.fna', '../../data_temp/Single-Family/Enterobacteriaceae/76.fna', '../../data_temp/Single-Family/Enterobacteriaceae/441.fna', '../../data_temp/Single-Family/Enterobacteriaceae/292.fna', '../../data_temp/Single-Family/Enterobacteriaceae/83.fna', '../../data_temp/Single-Family/Enterobacteriaceae/129.fna', '../../data_temp/Single-Family/Enterobacteriaceae/291.fna', '../../data_temp/Single-Family/Enterobacteriaceae/110.fna']\n",
      "Using previously computed OrthoANI values\n",
      "ITERATION 79 : kmer_length=10, c_val=200, canon=True\n",
      "Computing sketches ...\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[17], line 64\u001b[0m\n\u001b[1;32m     62\u001b[0m polycond \u001b[38;5;241m=\u001b[39m generate_polycond(\u001b[38;5;241m3\u001b[39m,\u001b[38;5;241m1000000007\u001b[39m,c_val,iteration)\n\u001b[1;32m     63\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mComputing sketches ...\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m---> 64\u001b[0m kmer_sketches \u001b[38;5;241m=\u001b[39m \u001b[43mcompute_kmer_sketches_parallel\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m     65\u001b[0m \u001b[43m    \u001b[49m\u001b[43mchecked_genome_files\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     66\u001b[0m \u001b[43m    \u001b[49m\u001b[43mkmer_length\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mkmer_length\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     67\u001b[0m \u001b[43m    \u001b[49m\u001b[43mkmer_class\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mkmer_class\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     68\u001b[0m \u001b[43m    \u001b[49m\u001b[43msketching_cond\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mpolycond\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     69\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcanonical\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mcanon\u001b[49m\n\u001b[1;32m     70\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     71\u001b[0m kmer_sketches_off1 \u001b[38;5;241m=\u001b[39m kmer_sketches[\u001b[38;5;241m1\u001b[39m:]\n\u001b[1;32m     72\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mSketching complete, computing ani ...\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "Cell \u001b[0;32mIn[12], line 9\u001b[0m, in \u001b[0;36mcompute_kmer_sketches_parallel\u001b[0;34m(genome_files, kmer_length, kmer_class, sketching_cond, canonical)\u001b[0m\n\u001b[1;32m      7\u001b[0m args \u001b[38;5;241m=\u001b[39m [(g,kmer_class,sketching_cond,kmer_length,canonical) \u001b[38;5;28;01mfor\u001b[39;00m g \u001b[38;5;129;01min\u001b[39;00m genome_files]\n\u001b[1;32m      8\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m Pool(max_pool) \u001b[38;5;28;01mas\u001b[39;00m p:\n\u001b[0;32m----> 9\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mp\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mstarmap\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcompute_kmer_sketches\u001b[49m\u001b[43m,\u001b[49m\u001b[43margs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/Desktop/UROP/GIS-2024/coding/ANI-est/lib/python3.12/site-packages/multiprocess/pool.py:375\u001b[0m, in \u001b[0;36mPool.starmap\u001b[0;34m(self, func, iterable, chunksize)\u001b[0m\n\u001b[1;32m    369\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mstarmap\u001b[39m(\u001b[38;5;28mself\u001b[39m, func, iterable, chunksize\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m):\n\u001b[1;32m    370\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m'''\u001b[39;00m\n\u001b[1;32m    371\u001b[0m \u001b[38;5;124;03m    Like `map()` method but the elements of the `iterable` are expected to\u001b[39;00m\n\u001b[1;32m    372\u001b[0m \u001b[38;5;124;03m    be iterables as well and will be unpacked as arguments. Hence\u001b[39;00m\n\u001b[1;32m    373\u001b[0m \u001b[38;5;124;03m    `func` and (a, b) becomes func(a, b).\u001b[39;00m\n\u001b[1;32m    374\u001b[0m \u001b[38;5;124;03m    '''\u001b[39;00m\n\u001b[0;32m--> 375\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_map_async\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfunc\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43miterable\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mstarmapstar\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mchunksize\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/Desktop/UROP/GIS-2024/coding/ANI-est/lib/python3.12/site-packages/multiprocess/pool.py:768\u001b[0m, in \u001b[0;36mApplyResult.get\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    767\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mget\u001b[39m(\u001b[38;5;28mself\u001b[39m, timeout\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m):\n\u001b[0;32m--> 768\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mwait\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtimeout\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    769\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mready():\n\u001b[1;32m    770\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mTimeoutError\u001b[39;00m\n",
      "File \u001b[0;32m~/Desktop/UROP/GIS-2024/coding/ANI-est/lib/python3.12/site-packages/multiprocess/pool.py:765\u001b[0m, in \u001b[0;36mApplyResult.wait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    764\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mwait\u001b[39m(\u001b[38;5;28mself\u001b[39m, timeout\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m):\n\u001b[0;32m--> 765\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_event\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mwait\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtimeout\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/usr/lib/python3.12/threading.py:655\u001b[0m, in \u001b[0;36mEvent.wait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    653\u001b[0m signaled \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_flag\n\u001b[1;32m    654\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m signaled:\n\u001b[0;32m--> 655\u001b[0m     signaled \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_cond\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mwait\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtimeout\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    656\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m signaled\n",
      "File \u001b[0;32m/usr/lib/python3.12/threading.py:355\u001b[0m, in \u001b[0;36mCondition.wait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    353\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:    \u001b[38;5;66;03m# restore state no matter what (e.g., KeyboardInterrupt)\u001b[39;00m\n\u001b[1;32m    354\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m timeout \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m--> 355\u001b[0m         \u001b[43mwaiter\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43macquire\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    356\u001b[0m         gotit \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n\u001b[1;32m    357\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "iteration = 0\n",
    "\n",
    "for analysis_type in analysis_types:\n",
    "    for analysis_data_dir in analysis_types[analysis_type]:\n",
    "        genome_files = get_genome_files(data_home_dir,analysis_type,analysis_data_dir)\n",
    "        checked_genome_files = check_genome_files(genome_files)\n",
    "\n",
    "        print(f\"{analysis_type=}, {analysis_data_dir=}\")\n",
    "        print(f\"Using genome files {checked_genome_files}\")\n",
    "        \n",
    "        checked_genome_files_off1 = checked_genome_files[1:]\n",
    "\n",
    "        ortho_ani_condition = (\n",
    "            ortho_ani_dataframe[\"genome_file_1\"].isin(checked_genome_files)\n",
    "        )\n",
    "\n",
    "        filtered_ortho_ani_dataframe = ortho_ani_dataframe[ortho_ani_condition]\n",
    "        complement_ortho_ani_dataframe = ortho_ani_dataframe[~ortho_ani_condition]\n",
    "\n",
    "        if len(filtered_ortho_ani_dataframe) == 0:\n",
    "            print(\"Computing OrthoANI values ...\")\n",
    "            ortho_ani_vals = compute_ortho_ani_parallel(checked_genome_files,checked_genome_files_off1)\n",
    "            print(\"Computed OrthoANI values\")\n",
    "        else:\n",
    "            print(\"Using previously computed OrthoANI values\")\n",
    "            ortho_ani_vals = filtered_ortho_ani_dataframe[\"orthoani_val\"]\n",
    "\n",
    "        computed_ortho_ani_dataframe = pd.DataFrame(\n",
    "            {\n",
    "                \"analysis_type\": analysis_type,\n",
    "                \"analysis_data_dir\": analysis_data_dir,\n",
    "                \"genome_file_1\": checked_genome_files[:-1],\n",
    "                \"genome_file_2\": checked_genome_files_off1,\n",
    "                \"orthoani_val\": ortho_ani_vals\n",
    "            }\n",
    "        )\n",
    "        ortho_ani_dataframe = pd.concat([complement_ortho_ani_dataframe,computed_ortho_ani_dataframe])\n",
    "\n",
    "        min_kmer_length, max_kmer_length = analysis_types[analysis_type][analysis_data_dir]\n",
    "        \n",
    "        for kmer_length in range(min_kmer_length,max_kmer_length+1):\n",
    "            for c_val in c_val_range:\n",
    "                canon = True\n",
    "\n",
    "                iteration += 1\n",
    "\n",
    "                print(f\"ITERATION {iteration} : {kmer_length=}, {c_val=}, {canon=}\")\n",
    "\n",
    "                checked_genome_files_off1 = checked_genome_files[1:]\n",
    "\n",
    "                kmer_sketch_ani_condition = (\n",
    "                    (kmer_sketch_dataframe[\"genome_file_1\"].isin(checked_genome_files)) &\n",
    "                    (kmer_sketch_dataframe[\"kmer_length\"] == kmer_length) &\n",
    "                    (kmer_sketch_dataframe[\"c_val\"] == c_val) &\n",
    "                    (kmer_sketch_dataframe[\"canonical\"] == canon)\n",
    "                )\n",
    "\n",
    "                filtered_kmer_sketch_ani_dataframe = kmer_sketch_dataframe[kmer_sketch_ani_condition]\n",
    "                complement_kmer_sketch_ani_dataframe = kmer_sketch_dataframe[~kmer_sketch_ani_condition]\n",
    "\n",
    "                if len(filtered_kmer_sketch_ani_dataframe) == 0:\n",
    "                    polycond = generate_polycond(3,1000000007,c_val,iteration)\n",
    "                    print(\"Computing sketches ...\")\n",
    "                    kmer_sketches = compute_kmer_sketches_parallel(\n",
    "                        checked_genome_files,\n",
    "                        kmer_length = kmer_length,\n",
    "                        kmer_class = kmer_class,\n",
    "                        sketching_cond = polycond,\n",
    "                        canonical = canon\n",
    "                    )\n",
    "                    kmer_sketches_off1 = kmer_sketches[1:]\n",
    "                    print(\"Sketching complete, computing ani ...\")\n",
    "                    kmer_ani_vals = compute_kmer_ani_parallel(kmer_sketches,kmer_sketches_off1)\n",
    "                    print(\"ANI computation complete\")\n",
    "                else:\n",
    "                    print(\"Using previously computed ANI values\")\n",
    "                    kmer_ani_vals = filtered_kmer_sketch_ani_dataframe[\"kmer_ani_val\"]\n",
    "\n",
    "\n",
    "                num_entries = len(kmer_ani_vals)\n",
    "                \n",
    "                computed_kmer_sketch_ani_dataframe = pd.DataFrame(\n",
    "                    {\n",
    "                        \"analysis_type\": analysis_type,\n",
    "                        \"analysis_data_dir\": analysis_data_dir,\n",
    "                        \"genome_file_1\": checked_genome_files[:-1],\n",
    "                        \"genome_file_2\": checked_genome_files_off1,\n",
    "                        \"kmer_length\": [kmer_length for _ in range(num_entries)],\n",
    "                        \"c_val\": [c_val for _ in range(num_entries)],\n",
    "                        \"canonical\": [canon for _ in range(num_entries)],\n",
    "                        \"kmer_ani_val\": kmer_ani_vals\n",
    "                    }\n",
    "                )\n",
    "\n",
    "\n",
    "                kmer_sketch_dataframe = pd.concat(\n",
    "                    [complement_kmer_sketch_ani_dataframe,computed_kmer_sketch_ani_dataframe]\n",
    "                )\n",
    "\n",
    "                correlation_condition = (\n",
    "                    (correlation_dataframe[\"analysis_type\"] == analysis_type) &\n",
    "                    (correlation_dataframe[\"analysis_data_dir\"] == analysis_data_dir) &\n",
    "                    (correlation_dataframe[\"kmer_length\"] == kmer_length) &\n",
    "                    (correlation_dataframe[\"c_val\"] == c_val) & \n",
    "                    (correlation_dataframe[\"canonical\"] == canon)\n",
    "                )\n",
    "\n",
    "                filtered_correlation_dataframe = correlation_dataframe[~correlation_condition] # remove to recompute\n",
    "\n",
    "                print(\"Computing correlations...\")\n",
    "\n",
    "\n",
    "                # filtered_ortho_ani_dataframe = ortho_ani_dataframe[ortho_ani_condition]\n",
    "                # filtered_kmer_sketch_ani_dataframe = kmer_sketch_dataframe[kmer_sketch_ani_condition]\n",
    "\n",
    "                # combined_ani_dataframe = filtered_ortho_ani_dataframe.merge(filtered_kmer_sketch_ani_dataframe)\n",
    "\n",
    "\n",
    "                # ortho_ani_vals = combined_ani_dataframe[\"orthoani_val\"]\n",
    "                # kmer_ani_vals = combined_ani_dataframe[\"kmer_ani_val\"]\n",
    "\n",
    "                \n",
    "                filter_o, filter_k = zip(*[(o,k) for o,k in zip(ortho_ani_vals,kmer_ani_vals) if (k > 0 and o > 0)])\n",
    "\n",
    "                dropped_zeros = len(ortho_ani_vals) - len(filter_o)\n",
    "                pearson_coeff = scipy.stats.pearsonr(filter_o,filter_k)\n",
    "                spearman_coeff = scipy.stats.spearmanr(filter_o,filter_k)\n",
    "\n",
    "                lin_fit = np.poly1d(np.polyfit(filter_o, filter_k, 1))\n",
    "                r2_val_raw = r2_score(filter_o,filter_k)\n",
    "                r2_val_lin = r2_score(filter_o,[lin_fit(o) for o in filter_o])\n",
    "\n",
    "\n",
    "\n",
    "                correlation_dataframe = pd.concat([filtered_correlation_dataframe,\n",
    "                    pd.DataFrame({\n",
    "                        \"analysis_type\": analysis_type,\n",
    "                        \"analysis_data_dir\": analysis_data_dir,\n",
    "                        \"kmer_length\": [kmer_length],\n",
    "                        \"c_val\": [c_val],\n",
    "                        \"canonical\": [canon],\n",
    "                        \"pearson_coeff\" : [pearson_coeff.statistic],\n",
    "                        \"pearson_coeff_pval\" : [pearson_coeff.pvalue],\n",
    "                        \"spearman_coeff\" : [spearman_coeff.statistic],\n",
    "                        \"spearman_coeff_pval\" : [spearman_coeff.pvalue],\n",
    "                        \"r2_val_raw\": [r2_val_raw],\n",
    "                        \"r2_val_lin\": [r2_val_lin],\n",
    "                        \"dropped_zeros\": dropped_zeros\n",
    "                    })],\n",
    "                )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(ortho_ani_dataframe)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(kmer_sketch_dataframe)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(correlation_dataframe)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ortho_ani_dataframe.to_csv(ortho_ani_filename)\n",
    "kmer_sketch_dataframe.to_csv(kmer_sketch_ani_filename)\n",
    "correlation_dataframe.to_csv(corrleation_filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def box_plot_OrthoANI(\n",
    "    analysis_type, \n",
    "    ortho_ani_dataframe\n",
    "):\n",
    "    plt.figure(figsize=(20,10),dpi=300)\n",
    "    \n",
    "    plt.title(f\"Box Plot of OrthoANI values for {analysis_type}\")\n",
    "    \n",
    "    \n",
    "    filtered_ortho_ani_df = ortho_ani_dataframe[ortho_ani_dataframe[\"analysis_type\"] == analysis_type]\n",
    "\n",
    "    plt.boxplot(\n",
    "        [filtered_ortho_ani_df[filtered_ortho_ani_df[\"analysis_data_dir\"] == analysis_data_dir][\"orthoani_val\"] for analysis_data_dir in analysis_types[analysis_type]],\n",
    "        labels=[analysis_data_dir for analysis_data_dir in analysis_types[analysis_type]],\n",
    "    )\n",
    "    plt.ylabel(\"OrthoAni Value\")\n",
    "\n",
    "    plt.savefig(f\"../../plots/correlations/{analysis_type}-OrthoANI-boxplot.png\")\n",
    "    plt.close()\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_correlation(\n",
    "        analysis_type, \n",
    "        correlation_dataframe,\n",
    "        c_val_range,\n",
    "        correlation_string,\n",
    "    ):\n",
    "\n",
    "    \n",
    "\n",
    "    plt.figure(figsize=(10,10),dpi=300)\n",
    "    plt.xlabel(\"K-mer Length\")\n",
    "    plt.ylabel(correlation_string)\n",
    "    plt.title(f\"{correlation_string} of {kmer_class.__name__} ANI Esimtation against K-mer Length on {analysis_type}\") \n",
    "    for analysis_data_dir in analysis_types[analysis_type]:\n",
    "        analysis_correlation_dataframe = correlation_dataframe[\n",
    "            (correlation_dataframe[\"analysis_type\"] == analysis_type) \n",
    "            & (correlation_dataframe[\"analysis_data_dir\"] == analysis_data_dir)\n",
    "        ]\n",
    "        min_kmer_length, max_kmer_length = analysis_types[analysis_type][analysis_data_dir]\n",
    "        kmer_lengths = [kmer_length for kmer_length in range(min_kmer_length,max_kmer_length+1)]\n",
    "        canonical = True\n",
    "\n",
    "        filtered_correlation_dataframe = analysis_correlation_dataframe[\n",
    "            analysis_correlation_dataframe[\"canonical\"] == canonical\n",
    "        ]\n",
    "        \n",
    "        \n",
    "        for c_val in c_val_range:\n",
    "            correlation = list(\n",
    "                filtered_correlation_dataframe[\n",
    "                    (filtered_correlation_dataframe[\"c_val\"] == c_val) &\n",
    "                    (filtered_correlation_dataframe[\"kmer_length\"].isin(set(kmer_lengths)))\n",
    "                ][correlation_string]\n",
    "            )\n",
    "            print(len(kmer_lengths))\n",
    "            print(len(correlation))\n",
    "            # plt.scatter(kmer_lengths,correlation,label=f\"{analysis_data_dir}\",s=[3 for _ in range(len(kmer_lengths))])\n",
    "            plt.plot(kmer_lengths,correlation,label=f\"{analysis_data_dir}\")\n",
    "\n",
    "    plt.legend()\n",
    "    plt.plot()\n",
    "    plt.savefig(f\"../../plots/correlations/{analysis_type}-{kmer_class.__name__}-{canonical=}-{correlation_string}.png\")\n",
    "    plt.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for analysis_type in analysis_types:\n",
    "    for correlation_string in [\"pearson_coeff\",\"spearman_coeff\",\"r2_val_raw\",\"r2_val_lin\",\"dropped_zeros\"]:\n",
    "        plot_correlation(\n",
    "            analysis_type=analysis_type,\n",
    "            correlation_dataframe=correlation_dataframe,\n",
    "            c_val_range=c_val_range,\n",
    "            correlation_string=correlation_string\n",
    "        )\n",
    "    \n",
    "    box_plot_OrthoANI(\n",
    "        analysis_type=analysis_type,\n",
    "        ortho_ani_dataframe=ortho_ani_dataframe\n",
    "    )\n",
    "\n",
    "    # analysis_correlation_dataframe = correlation_dataframe[\n",
    "    #     (correlation_dataframe[\"analysis_type\"] == analysis_type) &\n",
    "    #     (correlation_dataframe[\"analysis_data_dir\"] == analysis_data_dir)\n",
    "    # ]\n",
    "\n",
    "    # for canonical in (True,):\n",
    "    #     filtered_correlation_dataframe = analysis_correlation_dataframe[\n",
    "    #         analysis_correlation_dataframe[\"canonical\"] == canonical\n",
    "    #     ]\n",
    "    #     plt.figure(figsize=(10,10),dpi=300)\n",
    "    #     plt.xlabel(\"K-mer Length\")\n",
    "    #     plt.ylabel(\"Pearson Correlation\")\n",
    "    #     plt.title(\"Pearson Correlation of FracMinHash ANI Esimtation against K-mer Length at different c-value thresholds\")\n",
    "\n",
    "    #     for c_val in c_val_range:\n",
    "    #         pearson_correlation = list(\n",
    "    #             filtered_correlation_dataframe[\n",
    "    #                 (filtered_correlation_dataframe[\"c_val\"] == c_val) &\n",
    "    #                 (filtered_correlation_dataframe[\"kmer_length\"].isin(set(range(min_kmer_length,max_kmer_length+1))))\n",
    "    #             ][\"pearson_coeff\"]\n",
    "    #         )\n",
    "    #         print(len(kmer_lengths))\n",
    "    #         print(len(pearson_correlation))\n",
    "    #         plt.scatter(kmer_lengths,pearson_correlation,label=f\"c = {c_val}\")\n",
    "\n",
    "    #     plt.legend()\n",
    "    #     plt.plot()\n",
    "    #     plt.savefig(f\"../../plots/correlations/{analysis_type}-{analysis_data_dir}-{kmer_class.__name__}-{canonical=}-pearson.png\")\n",
    "    #     plt.close()\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    #     plt.figure(figsize=(10,10),dpi=300)\n",
    "    #     plt.xlabel(\"K-mer Length\")\n",
    "    #     plt.ylabel(\"Spearman Correlation\")\n",
    "    #     plt.title(\"Spearman Correlation of FracMinHash ANI Esimtation against K-mer Length at different c-value thresholds\")\n",
    "\n",
    "    #     for c_val in c_val_range:\n",
    "    #         spearman_correlation = [ \n",
    "    #             filtered_correlation_dataframe[\n",
    "    #                 (filtered_correlation_dataframe[\"c_val\"] == c_val) &\n",
    "    #                 (filtered_correlation_dataframe[\"kmer_length\"] == kmer_length)\n",
    "    #             ][\"spearman_coeff\"]\n",
    "    #             for kmer_length in range(min_kmer_length,max_kmer_length+1)\n",
    "    #         ]\n",
    "    #         print(kmer_lengths)\n",
    "    #         print(spearman_correlation)\n",
    "    #         plt.scatter(kmer_lengths,spearman_correlation,label=f\"c = {c_val}\")\n",
    "\n",
    "    #     plt.legend()\n",
    "    #     plt.plot()\n",
    "    #     plt.savefig(f\"../../plots/correlations/{analysis_type}-{analysis_data_dir}-{kmer_class.__name__}-{canonical=}-spearman.png\")\n",
    "    #     plt.close()\n",
    "\n",
    "            \n",
    "\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import imageio\n",
    "\n",
    "def plot_single_kmer(\n",
    "        ortho_vals_df,\n",
    "        kmer_vals_df,\n",
    "        kmer_length,\n",
    "):\n",
    "    \n",
    "    Path(f\"../../plots/temp/{analysis_type}-{analysis_data_dir}/\").mkdir(parents=True, exist_ok=True)\n",
    "    combined_df = pd.merge(ortho_vals_df,kmer_vals_df)\n",
    "    ortho_vals = combined_df[\"orthoani_val\"]\n",
    "    kmer_vals = combined_df[\"kmer_ani_val\"]\n",
    "    fig = plt.figure(figsize=(10,10),dpi=300)\n",
    "    ax = fig.add_subplot()\n",
    "    filter_o, filter_k = zip(*[(o,k) for o,k in zip(ortho_vals,kmer_vals) if (k > 0 and o > 0)])\n",
    "    lin_fit = np.poly1d(np.polyfit(filter_o, filter_k, 1))\n",
    "    ax.plot(filter_o,filter_o,\"r-\")\n",
    "    ax.scatter(filter_o,filter_k,marker=\"^\",s=[2 for _ in filter_o])\n",
    "    # ax.scatter(filter_o,filter_o,marker=\"o\",s=[2 for _ in filter_o])\n",
    "    ax.plot(np.unique(filter_o), lin_fit(np.unique(filter_o)))\n",
    "    plt.figtext(0.7,0.15,f\"Zeros dropped : {len(ortho_vals) - len(filter_o)}\")\n",
    "    pearson_coeff = scipy.stats.pearsonr(filter_o,filter_k)\n",
    "\n",
    "    plt.figtext(0.7,0.17,f\"Pearson Corr. Coeff.: {pearson_coeff.statistic:.3f}\")\n",
    "    plt.figtext(0.7,0.19,f\"Pearson Corr. p.: {pearson_coeff.pvalue:.3e}\")\n",
    "\n",
    "\n",
    "\n",
    "    ax.legend([\"OrthoANI\",f\"{kmer_length}-mer dots\",f\"{kmer_length}-mer line : {lin_fit}\"])\n",
    "\n",
    "    plotname = f\"../../plots/temp/{analysis_type}-{analysis_data_dir}/{analysis_type}-{analysis_data_dir}-{kmer_length}mer-estimated-ANI.png\"\n",
    "    plt.savefig(plotname)\n",
    "    plt.close()\n",
    "    return plotname"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for analysis_type in analysis_types:\n",
    "    for analysis_data_dir in analysis_types[analysis_type]:\n",
    "        Path(f\"../../plots/{analysis_type}-{analysis_data_dir}/\").mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "        analysis_images = []\n",
    "        min_kmer_length, max_kmer_length = analysis_types[analysis_type][analysis_data_dir]\n",
    "\n",
    "        for kmer_length in range(min_kmer_length,max_kmer_length+1):\n",
    "            for c_val in c_val_range:\n",
    "                canon = True\n",
    "                analysis_images.append(\n",
    "                    imageio.v3.imread(\n",
    "                        plot_single_kmer(\n",
    "                            ortho_vals_df = ortho_ani_dataframe[\n",
    "                                (ortho_ani_dataframe[\"analysis_type\"] == analysis_type)\n",
    "                                & (ortho_ani_dataframe[\"analysis_data_dir\"] == analysis_data_dir)\n",
    "                            ],\n",
    "                            kmer_vals_df = kmer_sketch_dataframe[\n",
    "                                (kmer_sketch_dataframe[\"analysis_type\"] == analysis_type)\n",
    "                                & (kmer_sketch_dataframe[\"analysis_data_dir\"] == analysis_data_dir)\n",
    "                                & (kmer_sketch_dataframe[\"kmer_length\"] == kmer_length)\n",
    "                                & (kmer_sketch_dataframe[\"c_val\"] == c_val)\n",
    "                                & (kmer_sketch_dataframe[\"canonical\"] == canon)\n",
    "                            ],\n",
    "                            kmer_length= kmer_length,\n",
    "                        )\n",
    "                    )\n",
    "                )\n",
    "        \n",
    "        print(len(analysis_images))\n",
    "        plot_gif_filename = f\"../../plots/{analysis_type}-{analysis_data_dir}/{analysis_type}-{analysis_data_dir}-estimated-ANI.gif\"\n",
    "        imageio.mimsave(plot_gif_filename,analysis_images,duration=0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ANI-est",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
